arroyo/processing/strategies/routing_producer.py

```diff
@@ -0,0 +1,156 @@
+import logging
+import time
+from abc import ABC, abstractmethod
+from collections import deque
+from concurrent.futures import Future
+from typing import Deque, MutableMapping, NamedTuple, Optional, Tuple
+
+from arroyo.backends.abstract import Producer
+from arroyo.backends.kafka import KafkaPayload
+from arroyo.processing.strategies import ProcessingStrategy
+from arroyo.types import Commit, Message, Partition, Position, Topic
+
+logger = logging.getLogger(__name__)
+
+
+class MessageRoute(NamedTuple):
+    """
+    A MessageRoute is a tuple of (producer, topic) that a message should be
+    routed to.
+    """
+
+    producer: Producer[KafkaPayload]
+    topic: Topic
+
+
+class MessageRouter(ABC):
```

Comment:
This class contains quite a lot of boilerplate for my taste. I don't think it's all needed. I would actually prefer if this was essentially just the `get_route_for_message` function, ie. we just pass a `Callalble[[Message[KafkaPayload]], MessageRoute]` to the strategy. What do you think?

Reply:
Originally I tried to build it the way you suggested. But the abstract base class keeps thing simpler to implement.
1. There are multiple interfaces which need implementation. See my comment above. If the router is holding onto some resources, it needs to release them for a clean shutdown.
2. If there is state which you would like to maintain in the router you can do that. IMO, callables are ideal for stateless behavior. But for things like round robin router etc, you might want to maintain state.

Reply:
Yeah, this is tightly related to the other comment, I'd prefer if shutdown wasn't the responsibility of the strategy at all.